### ISSUE TITLE
"Export type can only be built-in, a resource, or an enum" when @exporting a custom Resource

### ISSUE THREAD

**Issue by EIREXE** [godotengine/godot#49443]
### Godot version

4.0.dev (006cb7d)

### System information

Windows 10

### Issue description

When using a custom class with class_name and exporting it with `@export` this error is given:

![imagen](https://user-images.githubusercontent.com/3101690/121264887-2ed23d80-c8b8-11eb-803e-1b57e6ca7bbd.png)

### Steps to reproduce

Create two scripts with the following content:

BugTest.gd
```
extends Resource

class_name TestBug
```

BugTestNew.gd
```
extends Node

@export var bug_test := TestBug.new()
```

**Comment by pycbouh** [comment]
Is it a regression? You cannot export Nodes in 3.x neither, this isn't new.

**Comment by EIREXE** [comment]
> 
> 
> Is it a regression? You cannot export Nodes in 3.x neither, this isn't new.

Oh I made a mistake, the same issue happens if the original script is a Resource, I've ammended the original.

**Comment by EIREXE** [comment]
> 
> 
> TestBug in your example still doesn't extend Resource, it's a Node.

I shouldn't write while sleepy, the example is fixed now, still same problem (in fact I just made a mistake copying from the editor)

**Comment by ekumlin** [comment]
Is this issue the same as the request made by [proposal 18](https://github.com/godotengine/godot-proposals/issues/18)?

**Comment by pycbouh** [comment]
No, this issue is about a regression introduced in 4.0, not a missing feature

**Comment by ekumlin** [comment]
A regression against 3.3, or against a previous version of 4.0?

The reason I ask, and why I pointed to that proposal, is that the behavior of using `export` with a custom `Resource` type has never worked for me in any version of Godot 3.x. The error looks a little different in Godot 3.3, but it's there.

![image](https://user-images.githubusercontent.com/1350074/122699164-bba4c180-d1fd-11eb-8df1-456e65063b96.png)

If you use `export (X)`, then you get this:
![image](https://user-images.githubusercontent.com/1350074/122699221-d7a86300-d1fd-11eb-94be-793b3e0ee10c.png)
![image](https://user-images.githubusercontent.com/1350074/122699229-daa35380-d1fd-11eb-903d-11fae5f6c226.png)

If you use `export var foo: X` then you get this:
![image](https://user-images.githubusercontent.com/1350074/122699414-3a99fa00-d1fe-11eb-83b6-b86016d82d21.png)
![image](https://user-images.githubusercontent.com/1350074/122699423-3d94ea80-d1fe-11eb-8a3d-746a4f5bcc8d.png)

This leads me to believe this isn't actually a regression. What version did it last work with?

**Comment by EIREXE** [comment]
> 
> 
> A regression against 3.3, or against a previous version of 4.0?
> 
> The reason I ask, and why I pointed to that proposal, is that the behavior of using `export` with a custom `Resource` type has never worked for me in any version of Godot 3.x. The error looks a little different in Godot 3.3, but it's there.
> 
> ![image](https://user-images.githubusercontent.com/1350074/122699164-bba4c180-d1fd-11eb-8df1-456e65063b96.png)
> 
> If you use `export (X)`, then you get this:
> ![image](https://user-images.githubusercontent.com/1350074/122699221-d7a86300-d1fd-11eb-94be-793b3e0ee10c.png)
> ![image](https://user-images.githubusercontent.com/1350074/122699229-daa35380-d1fd-11eb-903d-11fae5f6c226.png)
> 
> If you use `export var foo: X` then you get this:
> ![image](https://user-images.githubusercontent.com/1350074/122699414-3a99fa00-d1fe-11eb-83b6-b86016d82d21.png)
> ![image](https://user-images.githubusercontent.com/1350074/122699423-3d94ea80-d1fe-11eb-8a3d-746a4f5bcc8d.png)
> 
> This leads me to believe this isn't actually a regression. What version did it last work with?

I don't know if it's necessarily a regression, but an issue nonetheless, it does say that resources (not just native ones) are allowed, so it not working would be a bug, either the text is wrong or the interpreter is. I personally think it's an issue.

**Comment by pycbouh** [comment]
@ekumlin Yes, you are correct with all that. In 3.x type inference doesn't work on this either, and mixing your custom type with a Resource-typed export hint also doesn't work. It only works when you force either the variable type, or the hint type to be a generic Resource (or any other built-in resource for that matter).

```
export var bug_test := TestResource.new() # Type-less export needs a constant expression assigned to infer type.
export(Resource) var bug_test : TestResource = TestResource.new() # The export hint's type (Resource) doesn't match the variable's type (TestResource).

export(Resource) var bug_test = TestResource.new() # OK
export var bug_test : Resource = TestResource.new() # OK
```

I guess the change in 4.0 where export hints no longer accept their own type and rely on the type system didn't change much here, and this is indeed more related to the proposal. We can keep this issue open in case we want to change the error message, as indeed it used to refer to the "native resource types", and doesn't now.

**Comment by vnen** [comment]
This isn't a bug, it's a lack of overall support for exporting custom resources (it goes beyond GDScript). Closing this in favor of https://github.com/godotengine/godot-proposals/issues/18 which will be implemented for Godot 4.0.
